### ISSUE TITLE
No way to get the latest reading/event data?

### ISSUE THREAD

**Issue by roychen11232357** [edgexfoundry/edgex-go#929]
Hi, there

I need to get the latest reading/event data via core-data API (https://docs.edgexfoundry.org/core-data.html#),

But It seems no way to filter out the latest reading/event data?

**Comment by jpwhitemn** [comment]
The export services are the means to get data from EdgeX to an endpoint of your choice (HTTP, MQTT, Azure IoT, Google IoT Core, etc.) as the data comes in (i.e. the latest events/reading).  Additionally, the export services offer a filter capability to reduce was is sent to you.  Today, the export services offer a simple way to filter by device or value descriptor associated to the readings (just include these filters as part of the export client registration).  One could easily modify the export distro "pipe and filter" architecture to incorporate additional filters.  See here for more details on export registration:  https://docs.edgexfoundry.org/Ch-APIExportClientRegistration-MQTT.html and here for generals on export services:  https://docs.edgexfoundry.org/Ch-ExportServices.html

**Comment by roychen11232357** [comment]
Dear jpwhitemn, 

Thanks for your response, but I already know the MQTT as endpoint the get export reading data from EdgeX, and also works fine using this way.

But if I want to use HTTP to get the latest reading, it seems no way currently?

I need the API such as:

`GET /reading/device/{deviceId}/{limit}/desc`

Thanks again for your patience and response.

**Comment by jduranf** [comment]
Why not using this?
`GET /event/{start}/{end}/{limit}`

**Comment by jpwhitemn** [comment]
@roychen11232357 - Roy, couple of observations... as Joan has suggested in the comments, if you desire, there is a rich API to make direct request of Events (and associated Readings) from Core Data based on a start and end time parameters or other inputs (see https://docs.edgexfoundry.org/core-data.html) .  Having said that, I would also advise caution in that in so doing you are engaging Core Data in a way that could be detrimental to system performance and potential changes in the future.  Core Data is meant to get data from south side as quickly as possible and asynchronously deliver that to the export side for use by other applications.  When you request core data to provide you that data, and depending on how often you make the request and which database may be in support of EdgeX, you could potential cause Core Data to spend more resources answering your requests versus addressing data ingestion.  Just something to consider as you architect your solution.

**Comment by jpwhitemn** [comment]
@roychen11232357 - the Core Data API is meant to provide a simple means to query the edge repository of sensor data for small things (say for example a use case where a rules engine wants to verify if an abnormal temperature reading has been seen in the last hour).  The export services are setup to sent the data to an MQTT topic, an HTTP (or HTTPS) end point, or some of the cloud IoT MQTT topics (as in Azure).  Again, please , have a look at the Export Service documentation for more details (https://docs.edgexfoundry.org/Ch-ClientRegistration.html, https://docs.edgexfoundry.org/Ch-Distribution.html)

**Comment by mhall119** [comment]
@roychen11232357 you can think of the Core Data API as a "pull" API that you or other EdgeX services can use to get data on demand from Core Data. It's intended for inter-service communication, but you can use it for to manually read data too.

The export services (HTTP, MQTT and others) are "push" APIs, they will automatically send new readings to your remote services (using HTTP or MQTT respectively) without you having to initiate it. You can optionally specify filters and transformations (compression, encryption, etc) on the data before it's exported, something you can't do with the Core Data API.

So, in summary, you can use the Core Data API for working with data at the edge, but use the Export services to get data off the edge and into your datacenter or cloud.
